src/twisted/conch/ssh/transport.py

```diff
@@ -637,83 +659,186 @@ def sendPacket(self, messageType, payload):
+        payloadSize = len(payload) + 1
+        isETM = self.currentEncryptions.isOutMACETM()
+
+        if not isETM:
+            # For not ETM, the packet length is also included.
+            payloadSize += 4
+
         bs = self.currentEncryptions.encBlockSize
-        # 4 for the packet length and 1 for the padding length
-        totalSize = 5 + len(payload)
-        lenPad = bs - (totalSize % bs)
+        lenPad = bs - (payloadSize % bs)
         if lenPad < 4:
             lenPad = lenPad + bs
-        packet = (
-            struct.pack("!LB", totalSize + lenPad - 4, lenPad)
-            + payload
-            + randbytes.secureRandom(lenPad)
+
+        # Create the unencrypted packet.
+        payloadPacket = (
+            struct.pack("!B", lenPad) + payload + randbytes.secureRandom(lenPad)
+        )
+
+        if isETM:
+            # For Encrypt-then-MAC the total size is not encrypted.
+            encryptedPayload = self.currentEncryptions.encrypt(payloadPacket)
+            totalSizePacket = struct.pack("!L", len(encryptedPayload))
+            macTarget = totalSizePacket + encryptedPayload
+            macPacket = self.currentEncryptions.makeMAC(
+                self.outgoingPacketSequence, macTarget
+            )
+            return macTarget + macPacket
+
+        # For non ETM, we subtract the 4 bytes added to compute the
+        # padding size.
+        # But we also encrypt the package length.
+        totalSizePacket = struct.pack("!L", payloadSize + lenPad - 4)
+        encryptedPacket = self.currentEncryptions.encrypt(
+            totalSizePacket + payloadPacket
+        )
+        macTarget = totalSizePacket + payloadPacket
+
+        macPacket = self.currentEncryptions.makeMAC(
+            self.outgoingPacketSequence, macTarget
         )
-        encPacket = self.currentEncryptions.encrypt(
-            packet
-        ) + self.currentEncryptions.makeMAC(self.outgoingPacketSequence, packet)
-        self.transport.write(encPacket)
-        self.outgoingPacketSequence += 1
 
-    def getPacket(self):
+        return encryptedPacket + macPacket
+
+    def getPacket(self) -> bytes | None:
         """
         Try to return a decrypted, authenticated, and decompressed packet
         out of the buffer.  If there is not enough data, return None.
 
-        @rtype: L{str} or L{None}
-        @return: The decoded packet, if any.
+        @return: The decoded packet, if any. C{None} is returned if no packet was received yet.
         """
-        bs = self.currentEncryptions.decBlockSize
-        ms = self.currentEncryptions.verifyDigestSize
-        if len(self.buf) < bs:
+        if len(self.buf) < self.currentEncryptions.decBlockSize:
             # Not enough data for a block
             return
+
+        try:
+            if self.currentEncryptions.isInMACETM():
+                payload = self._getPacketETM()
+            else:
+                payload = self._getPacketMTE()
```

Comment:
I think it would be a lot easier to read if these `...MTE` / `...ETM` methods were encapsulated within their own object, so there were a protocol for the operation, then a MTE implementation and an ETM implementation, then a call to `self._packetGetter.getPacket()` (or something) rather than switching to one of two mostly-similar methods.

Reply:
If you *are* going to make this a conditional though, there should still only be one assignment:
```suggestion
            payload = self._getPacketETM() if self.currentEncryptions.isInMACETM() else self._getPacketMTE()
```

Reply:
Thanks.

I will see what we can do... in the future, we will also have an AES-GCM implementation.

so inline `if / else` is not ok for the long term.

I went with these simple methods inside the SSHTranport as I don't feel that is that much code to move into an object.